aidial_adapter_openai/embeddings/azure_ai_vision.py

```diff
@@ -0,0 +1,153 @@
+from aidial_sdk.embeddings.response import Embedding, EmbeddingResponse, Usage
+from pydantic import BaseModel
+
+from aidial_adapter_openai.dial_api.embedding_inputs import (
+    collect_embedding_inputs,
+)
+from aidial_adapter_openai.dial_api.resource import AttachmentResource
+from aidial_adapter_openai.dial_api.storage import FileStorage
+from aidial_adapter_openai.utils.auth import OpenAICreds
+from aidial_adapter_openai.utils.http_client import get_http_client
+from aidial_adapter_openai.utils.resource import Resource
+
+# The latest Image Analysis API offers two models:
+# * version 2023-04-15 which supports text search in many languages,
+# * the legacy 2022-04-11 model which supports only English.
+_VERSION_PARAMS = {
+    "api-version": "2024-02-01",
+    "model-version": "2023-04-15",
+}
+
+
+def _get_auth_headers(creds: OpenAICreds) -> dict[str, str]:
+    if "api_key" in creds:
+        return {"Ocp-Apim-Subscription-Key": creds["api_key"]}
+
+    if "azure_ad_token" in creds:
+        return {"Authorization": f"Bearer {creds['azure_ad_token']}"}
+
+    raise ValueError("Invalid credentials")
+
+
+class VectorizeResponse(BaseModel):
+    class Config:
+        extra = "allow"
+
+    vector: List[float]
+
+
+async def embeddings(
+    creds: OpenAICreds,
+    deployment: str,
+    endpoint: str,
+    file_storage: FileStorage | None,
+    data: dict,
+) -> EmbeddingResponse:
+    input = EmbeddingsRequest.parse_obj(data)
+
+    async def on_text(text: str) -> str:
+        return text
+
+    async def on_attachment(attachment: Attachment) -> Resource:
+        return await AttachmentResource(attachment=attachment).download(
+            file_storage
+        )
+
+    inputs_iter: AsyncIterator[str | Resource] = collect_embedding_inputs(
+        input,
+        on_text=on_text,
+        on_attachment=on_attachment,
+    )
+
+    inputs: List[str | Resource] = [input async for input in inputs_iter]
+
+    async def _get_embedding(input: str | Resource) -> VectorizeResponse:
+        if isinstance(input, str):
+            return await _get_text_embedding(creds, endpoint, input)
+        elif isinstance(input, Resource):
+            return await _get_image_embedding(creds, endpoint, input)
+        else:
+            assert_never(input)
+
+    tasks: List[asyncio.Task[VectorizeResponse]] = [
+        asyncio.create_task(_get_embedding(input)) for input in inputs
+    ]
+
+    responses = await asyncio.gather(*tasks)
+    vectors = [
+        Embedding(embedding=r.vector, index=idx)
+        for idx, r in enumerate(responses)
+    ]
+
+    n = len(vectors)
+    usage = Usage(prompt_tokens=n, total_tokens=n)
+
+    return EmbeddingResponse(model=deployment, data=vectors, usage=usage)
+
+
+async def _get_image_embedding(
+    creds: OpenAICreds, endpoint: str, resource: Resource
+) -> VectorizeResponse:
+    resp = await get_http_client().post(
+        url=endpoint.rstrip("/") + "/computervision/retrieval:vectorizeImage",
```

Comment:
Since that's upstream endpoint, maybe let's pass the whole endpoint into upstream config, with `/computervision/...`?

Reply:
We need to combine `:vectorizeImage` and `:vectorizeText` endpoints into a single endpoint.

The valid options are one of the common prefixes of these two endpoints:

```
SERVICE_HOSTNAME/computervision/retrieval:vectorize
SERVICE_HOSTNAME/computervision/retrieval
SERVICE_HOSTNAME/computervision/
SERVICE_HOSTNAME
```

Since we could not use an endpoint which will actually be called, it's strange for me to ask the user to provide any half-baked URL, which is not documented in any way in the Azure docs.

The only reasonable and simplest option - is to provide the URL of the service itself as it's referenced under the name VISION_ENDPOINT in all of the docs about Image Analysis e.g. this [one](https://learn.microsoft.com/en-us/azure/ai-services/computer-vision/quickstarts-sdk/image-analysis-client-library-40?tabs=visual-studio%2Clinux&pivots=programming-language-python).